File: src/Utility/Crc/CrcGeneric.java
package Utility.Crc;

import Utility.BitMirror.BitMirror;

import java.util.zip.Checksum;

/**
 * Implements the Checksum interface as defined by java.util.zip.Checksum (the same
 * interface that CRC32 uses).
 *
 * http://www.sunshine2k.de/articles/coding/crc/understanding_crc.html#ch4 has great
 * examples on the theory behind calculating CRC values
 *
 * @author          Geoffrey Hunter <gbmhunter@gmail.com> (www.mbedded.ninja)
 * @since           2016-06-30
 * @last-modified   2016-06-30
 */
public class CrcGeneric implements Checksum {

    private static final int DATA_WIDTH_BITS = 8;

    private Integer crcWidthBits;
    private long crcPolynomial;
    private long startingValue;
    private Boolean reflectData;
    private long finalXorValue;
    private Boolean reflectRemainder;

    private long mask;
    private long crcValue;

    //================================================================================================//
    //========================================= CONSTRUCTORS =========================================//
    //================================================================================================//

    public CrcGeneric(
            Integer crcWidthBits,
            Long crcPolynomial,
            Long startingValue,
            Long finalXorValue,
            Boolean reflectData,
            Boolean reflectRemainder) {

        this.crcWidthBits = crcWidthBits;
        this.crcPolynomial = crcPolynomial;
        this.startingValue = startingValue;
        this.finalXorValue = finalXorValue;
        this.reflectData = reflectData;
        this.reflectRemainder = reflectRemainder;

        // Create a mask for future use in the update() method.
        // If the generator polynomial width is 8 bits, then the mask needs to be 0xFF,
        // if it is 16bits, then the mask needs to be 0xFFFF, e.t.c
        Double tempVal = Math.pow(2, crcWidthBits) - 1;
        mask = (long)tempVal.doubleValue();

        // Initialise the CRC value with the starting value
        crcValue = startingValue;

    }

    /**
     * Create a CRC engine based of a predefined CRC algorithm.
     * @param crcAlgorithmParameters
     */
    public CrcGeneric(CrcAlgorithmParameters crcAlgorithmParameters) {

        this.crcWidthBits = crcAlgorithmParameters.crcWidthBits;
        this.crcPolynomial = crcAlgorithmParameters.crcPolynomial;
        this.startingValue = crcAlgorithmParameters.startingValue;
        this.finalXorValue = crcAlgorithmParameters.finalXorValue;
        this.reflectData = crcAlgorithmParameters.reflectData;
        this.reflectRemainder = crcAlgorithmParameters.reflectRemainder;

        // Create a mask for future use in the update() method.
        // If the generator polynomial width is 8 bits, then the mask needs to be 0xFF,
        // if it is 16bits, then the mask needs to be 0xFFFF, e.t.c
        Double tempVal = Math.pow(2, crcWidthBits) - 1;
        mask = (long)tempVal.doubleValue();

        // Initialise the CRC value with the starting value
        crcValue = startingValue;

    }




    @Override
    public void update(int b) {

        long input = b;

        if(reflectData) {
            input = BitMirror.doMirror(input, 8);
        }

        // XOR-in the next byte of data, shifting it first
        // depending on the polynomial width.
        // This trick allows us to operate on one byte of data at a time before
        // considering the next
        crcValue ^= (input << (crcWidthBits - DATA_WIDTH_BITS));

        for (int j = 0; j < DATA_WIDTH_BITS; j++)
        {
            // Check to see if MSB is 1, if so, we need
            // to XOR with polynomial
            if ((crcValue & (1 << (crcWidthBits - 1))) != 0)
            {
                crcValue = ((crcValue << 1) ^ crcPolynomial) & mask;
            }
            else
            {
                crcValue = (crcValue << 1) & mask;
            }
        }
    }

    @Override
    public void update(byte[] b, int off, int len) {

    }

    @Override
    public long getValue() {

        long output = 0;
        if(reflectRemainder) {
            output = BitMirror.doMirror(crcValue, crcWidthBits);
        } else {
            output = crcValue;
        }

        output ^= finalXorValue;

        return output;
    }

    @Override
    public void reset() {
        crcValue = startingValue;
    }
}

File: src/misc/Crc/CrcGeneric.js
var bigInt = require('big-integer')

export class CrcGeneric {
  constructor (initObj) {
    console.log('CrcGeneric.initObj() called.')

    this.DATA_WIDTH_BITS = 8

    this.name = initObj.name
    this.crcWidthBits = initObj.crcWidthBits
    this.crcPolynomial = initObj.crcPolynomial
    this.startingValue = initObj.startingValue
    this.finalXorValue = initObj.finalXorValue
    this.reflectData = initObj.reflectData
    this.reflectRemainder = initObj.reflectRemainder
    this.checkValue = initObj.checkValue

    this.mask = bigInt('0')
    this.shiftedPolynomial = bigInt('0')

    // Create a mask for future use in the update() method.
    // If the generator polynomial width is 8 bits, then the mask needs to be 0xFF,
    // if it is 16bits, then the mask needs to be 0xFFFF, e.t.c
    var shiftingBit = bigInt('1')
    for (var i = 0; i < this.crcWidthBits; i++) {
      // mask |= shiftingBit;
      this.mask = this.mask.or(shiftingBit)
      // shiftingBit <<= 1;
      shiftingBit = shiftingBit.shiftLeft(1)
    }

    // this.shiftedPolynomial = (crcPolynomial << (8 - crcWidthBits));
    this.shiftedPolynomial = this.crcPolynomial.shiftLeft(8 - this.crcWidthBits)

    // Initialise the CRC value with the starting value
    this.crcValue = this.startingValue
  }

  update = (byteOfData) => {
    console.log('CrcGeneric.update() called.')

    var input = byteOfData

    if (this.reflectData) {
      input = this.doMirror(input, 8)
    }

    if (this.crcWidthBits - this.DATA_WIDTH_BITS >= 0) {
      // CRC POLYNOMIAL WIDTH >= DATA WIDTH

      // XOR-in the next byte of data, shifting it first
      // depending on the polynomial width.
      // This trick allows us to operate on one byte of data at a time before
      // considering the next
      this.crcValue ^= (input << (this.crcWidthBits - this.DATA_WIDTH_BITS))

      for (var j = 0; j < this.DATA_WIDTH_BITS; j++) {
        // Check to see if MSB is 1, if so, we need
        // to XOR with polynomial
        if ((this.crcValue & (1 << (this.crcWidthBits - 1))) !== 0) {
          this.crcValue = ((this.crcValue << 1) ^ this.crcPolynomial) & this.mask
        } else {
          this.crcValue = (this.crcValue << 1) & this.mask
        }
      }
    } else {
      // CRC POLYNOMIAL WIDTH < DATA WIDTH

      this.crcValue <<= this.DATA_WIDTH_BITS - this.crcWidthBits

      this.crcValue ^= input
      for (var k = 0; k < 8; k++) {
        this.crcValue = ((this.crcValue & 0x80) !== 0) ? (this.crcValue << 1) ^ this.shiftedPolynomial : this.crcValue << 1
      }

      this.crcValue &= 0xFF
      this.crcValue >>= this.DATA_WIDTH_BITS - this.crcWidthBits
    }
  }

  doMirror = (input, numBits) => {
    var output = 0

    for (var i = 0; i < numBits; i++) {
      output <<= 1
      output |= (input & 1)
      input >>= 1
    }

    return output
  }

}
